src/targets/gpu/device/topk.cpp

```diff
@@ -0,0 +1,225 @@
+            idxi[axis] = r;
+            idxr[axis] = ind[oss.index(idxi)];
+        }
+
+        if(l < n && op(arr[iss.index(idxl)], arr[iss.index(idx)]))
+        {
+            index = l;
+        }
+
+        if(r < n && op(arr[iss.index(idxr)], arr[iss.index(idx)]))
+        {
+            index = r;
+            if(op(arr[iss.index(idxl)], arr[iss.index(idxr)]))
+            {
+                index = l;
+            }
+        }
+
+        if(index == pre_index)
+        {
+            break;
+        }
+
+        idxi[axis] = index;
+        idxp[axis] = pre_index;
+        swap(ind[oss.index(idxi)], ind[oss.index(idxp)]);
+    }
+}
+
+template <class T, index_int N, class Op>
+__device__ inline void build_heap(T* arr,
+                                  int64_t* const ind,
+                                  const int64_t i,
+                                  const hip_shape<N>& oss,
+                                  const hip_shape<N>& iss,
+                                  const hip_shape<N>& css,
+                                  int n,
+                                  const int64_t axis,
+                                  Op op)
+{
+    for(int j = n / 2 - 1; j >= 0; j--)
+    {
+        heap_heapify(arr, ind, i, oss, iss, css, n, j, axis, op);
+    }
+}
+
+template <class T, index_int N, class Op>
+__device__ inline void heap_add(T* arr,
+                                int64_t* ind,
+                                const int64_t i,
+                                const hip_shape<N>& oss,
+                                const hip_shape<N>& iss,
+                                const hip_shape<N>& css,
+                                int n,
+                                int val,
+                                const int64_t axis,
+                                Op op)
+{
+    auto idx   = css.multi(i);
+    auto vidx  = idx;
+    idx[axis]  = ind[oss.index(idx)];
+    vidx[axis] = val;
+
+    if(op(arr[iss.index(vidx)], arr[iss.index(idx)]))
+    {
+        return;
+    }
+
+    idx                 = css.multi(i);
+    ind[oss.index(idx)] = val;
+    heap_heapify(arr, ind, i, oss, iss, css, n, 0, axis, op);
+}
+
+template <class T, index_int N, class Op>
+__device__ inline void heap_sort(T* arr,
+                                 int64_t* const ind,
+                                 const int64_t i,
+                                 const hip_shape<N>& oss,
+                                 const hip_shape<N>& iss,
+                                 const hip_shape<N>& css,
+                                 int n,
+                                 const int64_t axis,
+                                 Op op)
+{
+    build_heap(arr, ind, i, oss, iss, css, n, axis, op);
+    auto idx  = css.multi(i);
+    auto idxj = idx;
+    for(int j = n - 1; j > 0; --j)
+    {
+        idxj[axis] = j;
+        swap(ind[oss.index(idx)], ind[oss.index(idxj)]);
+        heap_heapify(arr, ind, i, oss, iss, css, j, 0, axis, op);
+    }
+}
```

Comment:
Is it possible to make a heap class to reduce the number of parameters that need to be passed to these functions?